Graphviz debug output for generic dataflow analysis

ecstatic-morse opened the pull request:
A follow up to #64566.

This outputs graphviz diagrams in the generic dataflow engine when `#[rustc_mir(borrowck_graphviz_postflow="suffix.dot")]` is set on an item. This code is based on [`dataflow/graphviz.rs`](https://github.com/rust-lang/rust/blob/master/src/librustc_mir/dataflow/graphviz.rs), but displays different information since there are no "gen"/"kill" sets and transfer functions cannot be coalesced in the generic analysis. As a result, we show the dataflow state at the start and end of each block, as well as the changes resulting from each statement. I also render the state bitset in full  (`{_1,_2}`) instead of hex-encoded as the current renderer does (`06`).

rust-highfive commented:
r? @zackmdavis

(rust_highfive has picked a reviewer for you, use r? to override)

ecstatic-morse commented:
r? @oli-obk

ecstatic-morse commented:
Oh, this should only get merged after #64470 does. There's some small conflicts now that `dataflow::generic::Engine::new` takes some additional parameters.

eddyb commented:
I think it would be neat to use [`debug_set`](https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_set) for more things, including all of our set abstractions, and this.

Instead of `(_1,_2)`, you'd get `{_1, _2}` which looks more like a set to me, at least.

ecstatic-morse commented:
The only reason I didn't go with `debug_set` was to get rid of the space after the comma. I'm considering changing to the hex representation (`14-ef`) when the cardinality crosses a certain threshold, which would make this less of an issue. I also thought about just printing the full set on entry and exit, and only printing the diff for each statement, but this is kind of a pain if you want to know the state in the middle of a longer basic blocks.

ecstatic-morse commented:
I switched to only rendering the diff for each statement but printing the full dataset at block entry and exit. This makes it a lot easier to track changes to the dataflow state when there are a lot of elements, but makes it more difficult to quickly see the whole state at a given statement. I also improved the formatting a bit. Here's the new on the same input file:

![prettier](https://user-images.githubusercontent.com/29463364/65812320-81b6a080-e17a-11e9-80a2-ed66bd414c96.png)

This particular analysis never clears dataflow bits, but those would appear in red prefixed with a minus.

bors commented:
:umbrella: The latest upstream changes (presumably #64778) made this pull request unmergeable. Please resolve the merge conflicts.

ecstatic-morse commented on src/librustc_mir/dataflow/generic.rs:
```diff
@@ -316,7 +338,7 @@ where
 
         let block_data = &self.body.basic_blocks()[target_block];
-        self.results.analysis.apply_partial_block_effect(
+        self.results.borrow().analysis.apply_partial_block_effect(
```
These won't pass the borrow checker since we need to mutably borrow `self.state` and pass it as an argument to `Analysis::apply_partial_block_effect`.

ecstatic-morse commented:
I've fixed merge conflicts and cleaned up the code a bit. I also added an extra row to the table for each basic block, so that both the diff for the terminator and the full state after the terminator's transfer function is applied are displayed. I think this is ready to go.

![prettiest](https://user-images.githubusercontent.com/29463364/65905366-4cc26d80-e375-11e9-9e55-acb6fbde9b46.png)

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 2b8e023b9d28f2f912ad21427b5266b62c007f11 has been approved by `oli-obk`

<!-- @bors r=oli-obk 2b8e023b9d28f2f912ad21427b5266b62c007f11 -->
<!-- homu: {"type":"Approved","sha":"2b8e023b9d28f2f912ad21427b5266b62c007f11","approver":"oli-obk"} -->
